sdk/core/switcher/entry.S

```diff
@@ -126,6 +135,46 @@ switcher_scheduler_entry_csp:
  * and integer register.  All three registers are clobbered.
  */
 .macro zero_stack base top scratch
+#ifdef CHERIOT_HAS_ZTOP
+	// Derive the capability to the range that should be zeroed.
+	// This is stored in base, leaving top as a second scratch register to use
+	csub               \top, c\top, c\base
+	csetboundsexact    c\base, c\base, \top
+
+	// Wait for any prior zeroing to finish.
+	cspecialr          c\scratch, SCR_ZTOP
+	// If ztop is untagged, no zeroing is happening and our attempt to store
+	// relative to ztop will trap.
+	cgettag            \top, c\scratch
+	beqz               \top, 2f
+	// If we've already reached the bottom then do nothing.
+	// NOTE: This can be deleted if ZTOP's tag is cleared when zeroing finishes.
+	cgetbase           \top, c\scratch
+	beq                \top, \scratch, 2f
+	// If the current zeroing range is a subset of range that we're about to
+	// zero, don't bother waiting, just zero from the start.
+	ctestsubset        \top, c\base, c\scratch
```

Comment:
We could do it with checks on the top and bottom, since we already have checks that CSP is sane (correct permissions, correctly aligned top and bottom) on entry into the compartment switcher.  I'm not sure if we could do that without adding another temporary register, which would require some small tweaks to the users of this macro.